### PR TITLE
[Merged by Bors] - feat: port SetTheory.ZFC.Basic

### PR DESCRIPTION
**Pull request by vihdzp**
A re-do of #1215, necessary since a lot of major changes have happened in the last months.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Zfc/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,1767 @@
...
+  ZFSet.sep (IsFunc x y) (powerset (prod x y))
+#align Set.funs ZFSet.funs
+
+@[simp]
+theorem mem_funs {x y f : ZFSet.{u}} : f ∈ funs x y ↔ IsFunc x y f := by simp [funs, IsFunc]
+#align Set.mem_funs ZFSet.mem_funs
+
+/- Porting note: the `Definable` argument in `mapDefinableAux` is unused, though the TODO remark
+   suggests it shouldn't be. -/
+set_option linter.unusedVariables false
```
Comment (vihdzp) [inline Mathlib/SetTheory/Zfc/Basic.lean:1337]: This doesn't actually disable the linter. Am I using the wrong option?

Comment (digama0) [reply Mathlib/SetTheory/Zfc/Basic.lean:1340]: There are actually two unused variable linters. One of them reports immediately on affected variables and is controlled by ` set_option linter.unusedVariables` (this is the lean core linter), and the other one reports on `#lint` and is controlled by `@[nolint unusedVariables]` (this is the mathlib linter).

File: Mathlib/SetTheory/Zfc/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1775 @@
...
+#align Set.mem_inter ZFSet.mem_inter
+
+@[simp]
+theorem mem_diff {x y z : ZFSet.{u}} : z ∈ x \ y ↔ z ∈ x ∧ z ∉ y :=
+  @mem_sep (fun z : ZFSet.{u} => z ∉ y) x z
+#align Set.mem_diff ZFSet.mem_diff
+
+@[simp]
+theorem unionₛ_pair {x y : ZFSet.{u}} : ⋃₀ ({x, y} : ZFSet.{u}) = x ∪ y :=
+  rfl
```
Comment (vihdzp) [inline Mathlib/SetTheory/Zfc/Basic.lean:1158]: The Mathlib 3 proof isn't `rfl` but could be, I'll fix that at a later time.

**Comment by vihdzp** [comment]
The file has been ported up to SHA [98bbc3526516bca903bff09ea10c4206bf079e6b](https://github.com/leanprover-community/mathlib/commit/98bbc3526516bca903bff09ea10c4206bf079e6b). I would like for this to be ported before I port any further Mathlib 3 changes.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**